packages/protocol/contracts/common/libraries/B12.sol

```diff
@@ -0,0 +1,895 @@
+pragma solidity ^0.5.13;
+
+import { TypedMemView } from "@summa-tx/memview.sol/contracts/TypedMemView.sol";
+
+library B12 {
+  using TypedMemView for bytes;
+  using TypedMemView for bytes29;
+
+  // Fp is a field element with the high-order part stored in `a`.
+  struct Fp {
+    uint256 a;
+    uint256 b;
+  }
+
+  // Fp2 is an extension field element with the coefficient of the
+  // quadratic non-residue stored in `b`, i.e. p = a + i * b
+  struct Fp2 {
+    Fp a;
+    Fp b;
+  }
+
+  // G1Point represents a point on BLS12-377 over Fp with coordinates (X,Y);
+  struct G1Point {
+    Fp X;
+    Fp Y;
+  }
+
+  // G2Point represents a point on BLS12-377 over Fp2 with coordinates (X,Y);
+  struct G2Point {
+    Fp2 X;
+    Fp2 Y;
+  }
+
+  struct G1MultiExpArg {
+    G1Point point;
+    uint256 scalar;
+  }
+
+  struct G2MultiExpArg {
+    G2Point point;
+    uint256 scalar;
+  }
+
+  struct PairingArg {
+    G1Point g1;
+    G2Point g2;
+  }
+
+  function FpEq(Fp memory a, Fp memory b) internal pure returns (bool) {
+    return (a.a == b.a && a.b == b.b);
+  }
+
+  function fpGt(Fp memory a, Fp memory b) internal pure returns (bool) {
+    return (a.a > b.a || (a.a == b.a && a.b > b.b));
+  }
+
+  function Fp2Eq(Fp2 memory a, Fp2 memory b) internal pure returns (bool) {
+    return FpEq(a.a, b.a) && FpEq(a.b, b.b);
+  }
+
+  function fp2Gt(Fp2 memory a, Fp2 memory b) internal pure returns (bool) {
+    if (FpEq(a.b, b.b)) return fpGt(a.a, b.a);
+    else return fpGt(a.b, b.b);
+  }
+
+  function fpAdd2(uint256 a, uint256 b) internal pure returns (Fp memory) {
+    return fpAdd(Fp(0, a), Fp(0, b));
+  }
+
+  function fpAdd3(uint256 a, uint256 b, uint256 c) internal pure returns (Fp memory) {
+    return fpAdd(Fp(0, a), fpAdd(Fp(0, b), Fp(0, c)));
+  }
+
+  function fpAdd4(uint256 a, uint256 b, uint256 c, uint256 d) internal pure returns (Fp memory) {
+    return fpAdd(Fp(0, a), fpAdd(Fp(0, b), fpAdd(Fp(0, c), Fp(0, d))));
+  }
+
+  function fpAdd(Fp memory a, Fp memory b) internal pure returns (Fp memory) {
+    uint256 bb = a.b + b.b;
+    uint256 aa = a.a + b.a + (bb >= a.b && bb >= b.b ? 0 : 1);
+    return Fp(aa, bb);
+  }
+
+  function fpSub(Fp memory a, Fp memory b) internal pure returns (Fp memory) {
+    Fp memory p = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+    );
```

Comment:
Don't have the expertise to say for sure, but just want to say this looks fishy to me and it'd be great to get someone that knows more about BLS crypto to look at this.

The number here appears to be the base field modulus for BLS377 (see [here](https://eips.ethereum.org/EIPS/eip-2539#specification) for example; different from the BLS381 `p` as seen e.g. [here](https://github.com/zkcrypto/bls12_381)).

It doesn't appear that this function ever (directly or indirectly) gets called by the BLS381 library, but this just points to some weird organization of these libraries. Seems like some of the BLS377-specific functions are defined in B12, some in the BLS377 library, while there doesn't seem to be any BLS381-specific functions in B12. It might be that the code is still all correct, but I'm afraid this sort of code organization could lead to future developer confusion (for cLabs or external devs relying on these wrappers).

Reply:
Looking a little more deeply, seems like all the `fp<operation>` functions are only used to implement `mapToG1` and `mapToG2` in B12, in an apparently BLS377-specific way. The BLS381 library implements these `mapTo` operations calling out to precompiles. [CIP30](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0030.md) (for BLS-377) did not add these "map to G1/2" precompiles, while [CIP31](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0031.md), for BLS-381, did.

Since the precompiles are not available for BLS-377, makes sense to implement them manually in Solidity, but I guess it would make more sense for these functions to exist in the BLS377 library, rather than the general B12.

Reply:
Some fp functions are used in BLS12Passthrough.sol directly from `B12` and not through one of the specialized libraries. I'm not sure if such a generic usage is a relevant use case outside of testing, but it is at least one reason against moving the fp functions.